### PR TITLE
<iomanip>: fixes get_time to parse without delimiters

### LINKED ISSUE #944: <iomanip>: get_time should parse format without delimiters
**Command-line test case**
```
#include <iostream>
#include <sstream>
#include <iomanip>
#include <assert.h>

int main()
{
  {
    // Parsing with delimiters (such as spaces) works fine.
    std::tm t = {};
    std::istringstream ss("2020 06 20");
    ss >> std::get_time(&t, "%Y %m %d");
    assert( !ss.fail() );
  }

  {
    // Parsing without delimiters fails, but should work too.
    std::tm t = {};
    std::istringstream ss("20200620");
    ss >> std::get_time(&t, "%Y%m%d");
    if( ss.fail() )
        std::cout << "test failure\n";
  }
}
```

**STL version**
    Microsoft Visual Studio Professional 2019
    Version 16.6.2

**Additional context**
The standard draft WG21-N4861 says that the conversion specification is valid for the ISO/IEC 9945 function strptime.

Looking at strptime, it says "The application shall ensure that there is white-space or other non-alphanumeric characters between any two conversion specifications unless all of the adjacent conversion specifications convert a known, fixed number of characters."

The number of characters for each conversion specifier is listed.
https://pubs.opengroup.org/onlinepubs/9699919799/functions/strptime.html

See also: DevCom-9505, DevCom-140618, DevCom-246365 and Microsoft-internal VSO-370548 / [AB#370548](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/370548).

### PR DESCRIPTION
**Pull request by Arzaghi**
Fixes #944

### CODE REVIEW

File: stl/inc/xloctime
Comment (StephanTLavavej) [inline stl/inc/xloctime:542]: I believe this will break binary compatibility because it's changing the signature of a separately compiled function. (Much iostreams code is separately compiled; a subtle indication is the explicit calling convention `__CLRCALL_OR_CDECL`.) `dumpbin /exports` and grepping for `_Getint` should reveal whether the signature is changing. (We have a build step to validate that the export surface is unchanged, but we still need to port it from the MSVC-internal MSBuild system to the GitHub CMake/Ninja build system - sorry!)

It may be possible to fix this while preserving bincompat (as @statementreply encountered and solved a similar problem). I'll have time after CppCon to investigate in detail although another maintainer may have time before then.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/xloctime
Comment (Arzaghi) [reply stl/inc/xloctime]: Thank you for the details. Hopefully, in this case, it was easy to find another solution without touching the function signature thanks to `_Hi` parameter.

File: stl/inc/xloctime
Comment (StephanTLavavej) [inline stl/inc/xloctime:569]: This logic for resetting the number of digits seen when leading zeros are encountered appears to be incorrect - it results in successfully parsing `istringstream iss("202000000000000923");`, but Clang on Compiler Explorer reports `ios_base::failbit` for that. Therefore, I recommend adding that as a test case.

(The existing line "replace one or more with single zero" appears to be fine, as it's used for the actual conversion.)

File: stl/inc/xloctime
Comment (StephanTLavavej) [inline stl/inc/xloctime:542]: Naming nitpick: The codebase isn't entirely consistent, but we generally use `_UglyCamelCase` for types (like `_InIt`, `_FwdItHaystack`, etc.) and `_Ugly_snake_case` for variables. Here, I think `_Hi_digits` would be fine.

File: stl/inc/xloctime
Comment (StephanTLavavej) [inline stl/inc/xloctime:561]: Similarly, `_Seen_digits` or better yet `_Digits_seen` would be more conventional/readable.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/xloctime
Comment (Arzaghi) [reply stl/inc/xloctime]: Nice catch. Sorry for missing this case.

File: stl/inc/xloctime
Comment (Arzaghi) [reply stl/inc/xloctime]: I will try to follow this convention on my next PRs.

**Comment by Arzaghi** [comment]
An irrelevant issue I encountered during fixing this issue. I generally use clang-format extension for Visual Studio and apply the .clang-format file using `Tools->Clang Format Document`. But this option format these lines as follow:

```cpp
        const int _Hi_digits = (_Hi <= 9      ? 1
                                : _Hi <= 99   ? 2
                                : _Hi <= 999  ? 3
                                : _Hi <= 9999 ? 4
                                              : static_cast<int>(_STD log10(_STD abs(_Hi))) + 1);
```

which looks very clear and easy to read.
But the clang-format validation on the sever reject it and expect those lines as follow:

https://github.com/microsoft/STL/blob/7613ad9eed0d057d84b40dbc1671f141c47d843e/stl/inc/xloctime#L542-L545


I use the latest version of Clang-Format and the VS extension on my system. I'm wondering about the root of this difference and I'm not sure whether something wrong with a configuration on my system or this happening because of the difference between Clang-Format versions on my system and the server.
By the way I encountered some other similar minor conflicts on my previous PRs.

File: stl/inc/xloctime
Comment (StephanTLavavej) [reply stl/inc/xloctime]: No problem, that's what code review is for :smile_cat:

**Comment by StephanTLavavej** [comment]
Different versions of clang-format do change behavior; we are currently using 10.0.0. (Failed clang-format checks will print out the version used; it's the one installed by the VS Preview that we're using.)

In theory, if you're using the same version, and you're using the `--style=file` option to pick up our `.clang-format` configuration, the results _should_ be deterministically identical.

File: stl/inc/xloctime
Comment (StephanTLavavej) [inline stl/inc/xloctime:545]: This code is correct, but upon looking at it again, I noticed that `_STD abs(_Hi)` is unnecessary, because negative values would already have been absorbed by `_Hi <= 9 ? 1`. Then I wondered whether we need to worry about negative or enormous values at all - and I believe we don't. This is an internal function, and we pass constants for all of the `_Lo` and `_Hi` values. `_Hi` is always positive, and at most `9999`:
https://github.com/microsoft/STL/blob/5f736efd868f0cc32dd4c89d43f8ad9923742647/stl/inc/xloctime#L357
Therefore, I think it would be simpler (and less codegen in debug mode) to reduce the logic here, and add `_STL_INTERNAL_CHECK`s (which are assertions activated only for the STL's test suite, never for user compilation) to detect maintenance issues in the future.

**Review by StephanTLavavej — COMMENTED** [review]
This looks good to me - I especially like the set of test cases you've crafted. I'll push a change to simplify the `_Hi_digits` computation and move this forward to Final Review :smile_cat:

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/xloctime
Comment (Arzaghi) [reply stl/inc/xloctime]: I didn't know about `_STL_INTERNAL_CHECK`. So, I tried to let `_Getint` be general and valid for possible future usages. Using such assertation is very reasonable here. :+1:

**Review by CaseyCarter — APPROVED** [review]
Impressively exhaustive tests!

**Comment by Arzaghi** [comment]
@StephanTLavavej @CaseyCarter 
I've added one more tiny (yet important) test case to have better coverage for this scenario.
Excuse me I'm adding this test case at this final stage and after your review and approval.

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks - both for the additional test case of trailing non-digit characters, and the heads up :smile_cat:

**Comment by StephanTLavavej** [comment]
Thanks again for fixing this widely reported runtime correctness bug! Your fix will ship in VS 2019 16.9 Preview 2. :cat2:

**Comment by randydu** [comment]
The VS 2019 Preview 3 (Visual Studio 2019 Developer Command Prompt v16.9.0-pre.3.0) is installed but the following code cannot parse iso-string properly:

```c
    std::tm tm{0};
    std::istringstream ss(str);
    ss >> std::get_time(&tm, "%Y%m%dT%H%M%S");
    if (ss.fail())
        throw std::invalid_argument("Invalid argument: " + str);

   //str = "19700101T000001"
```
Expected:

```c
  CHECK(tm.tm_sec == 1); //should be 1 second
  CHECK(tm.tm_min == 0);
```

Actual:
```c
  CHECK(tm.tm_sec == 0);
  CHECK(tm.tm_min == 1); // 1 min ?
```

The same test works fine for gcc and clang.

**Pull request merged**
